Support BagIt in transfers

sevein opened the pull request:
Update `_parse_mets` so baggited transfers are indexed properly.
It is backward compatible - should be able to handle non-baggited transfers.
It's been also refactored so it uses `metsrw`.

Connects to https://github.com/archivematica/Issues/issues/444.
Connects to https://github.com/archivematica/Issues/issues/446.

sevein committed:
Support BagIt in transfers

Update `_parse_mets` so baggited transfers are indexed properly.

It is backward compatible - should be able to handle non-baggited transfers.

sevein committed:
Bump version to 0.15.0

sevein commented on storage_service/locations/fixtures/tiny_transfer/metadata/submissionDocumentation/METS.xml:
```diff
@@ -0,0 +1,65 @@
+      </mets:div>
+    </mets:div>
+  </mets:structMap>
+</mets:mets>
```
This transfer is incomplete but carries enough information for the existing tests.

sevein commented on storage_service/locations/tests/test_package.py:
```diff
@@ -126,9 +129,9 @@ def test_fixity_failure(self):
-        assert message == 'invalid bag'
+        assert len(failures) == 1
+        assert isinstance(failures[0], bagit.FileMissing)
+        assert message == 'Bag validation failed'
```
The version bump of bagit==1.7.0 changed the way errors are returned.

cole commented on storage_service/locations/models/package.py:
```diff
@@ -1524,16 +1524,13 @@ def compress_package(self, algorithm, extract_path=None):
         return (compressed_filename, extract_path)
 
-    def _parse_mets(self, prefix=None, relative_path=['metadata', 'submissionDocumentation', 'METS.xml']):
+    def _parse_mets(self, prefix):
```
I think this is a good opportunity to move this function over to using `metsrw` rather than its own logic if possible. If `metsrw` doesn't work for for this case (and it should result in a simpler function) maybe we can make some improvements to it in future.

cole commented on storage_service/locations/models/package.py:
```diff
@@ -2841,3 +2842,12 @@ def write_pointer_file(pointer_file, pointer_file_path):
+def _is_bagit(path):
+    """Determine whether ``path`` is a BagIt package."""
+    try:
+        bagit.Bag(path)
```
👍 looking a the bagit package, this seems like this is a good test — it checks for all the files a bag needs without running full validation. One thing I didn't think about is it may be slow for large bagit manifests, but I think if that becomes a problem we can work to optimize it later.

sevein replied:
Good point. It fails pretty fast when it's not a BagIt, but it took more than a minute to open one with 500k files.

```
$ time python3 -c "import bagit; bagit.Bag('/home/jesus/archivematica-sampledata')"
60.18user 12.11system 1:12.73elapsed 99%CPU (0avgtext+0avgdata 441052maxresident)k
624inputs+0outputs (0major+140945minor)pagefaults 0swaps
```

cole commented on storage_service/locations/tests/test_package.py:
```diff
@@ -238,3 +241,57 @@ def test_extract_file_aip_from_compressed_aip(self):
+    def tearDown(self):
+        shutil.rmtree(self.tmp_dir)
+
+    def _rsync_fixture(self, fixture_dir, dst, make_bagit=False):
```
Does `shutil.copytree` work instead of rsync? It seems better to avoid a dependency on rsync if we can (I know it's in basically all distros, but it's not core, and it requires the use of subprocess).

sevein replied:
Cool, it works.

cole reviewed (commented):
Looks good overall!

sevein committed:
Use metsrw in Package._parse_mets

The new version of metsrw fixes some encoding problems and adds compatibility
with old-style fptrs like the ones in the current transfer METS in
`tiny_fixtures`.

sevein commented:
Ready for CR again. Now `Package._parse_mets` is using metsrw and its latest release v0.3.0.
SS is ready for baggited transfers!

FYI we can't merge until the issues get a milestone but we'll probably get that done on Monday.

sevein committed:
Avoid working tree writes in tests

This commit updates the tests in `locations` that were mutating the source
tree so they use temporary directories instead. They are created using a new
`TempDirMixin` that instantiates a `Path` object. `pathlib` is made available
by pytest via `pathlib2`.

sevein committed:
Location: disable post_save API key signal in tests

This avoids a conflict where our tests would fail to insert the API key when
the database was being reused because the signal receiver had already created
it.

sevein committed:
Fix integrity error reusing db in tests

Base `setUpClass` needs to be called first so the transaction is initialized
before we mutate the data.

This solves a conflic raised when using `--reuse-db`.

sevein committed:
Fix Travis build issues

* Ensure that Py3 tests are reaching the interpreter.
* Use Py2 in `checkformigrations`, it fails otherwise.

sevein committed:
Update metsrw to 0.3.1
